Cabal/Distribution/Simple/Glob.hs

```diff
@@ -0,0 +1,177 @@
+-- slash and backslash as its path separators, if we left in the
+-- separators from the glob we might not end up properly normalised.
+
+data GlobSyntaxError
+  = StarInDirectory
+  | StarInFileName
+  | StarInExtension
+  | NoExtensionOnStar
+  | EmptyGlob
+  | LiteralFileNameGlobStar
+  | VersionDoesNotSupportGlobStar
+  | VersionDoesNotSupportGlob
+  deriving (Eq, Show)
+
+explainGlobSyntaxError :: FilePath -> GlobSyntaxError -> String
+explainGlobSyntaxError filepath StarInDirectory =
+     "invalid file glob '" ++ filepath
+  ++ "'. A wildcard '**' is only allowed as the final parent"
+  ++ " directory. Stars must not otherwise appear in the parent"
+  ++ " directories."
+explainGlobSyntaxError filepath StarInExtension =
+     "invalid file glob '" ++ filepath
+  ++ "'. Wildcards '*' are only allowed as the"
+  ++ " file's base name, not in the file extension."
+explainGlobSyntaxError filepath StarInFileName =
+     "invalid file glob '" ++ filepath
+  ++ "'. Wildcards '*' may only totally replace the"
+  ++ " file's base name, not only parts of it."
+explainGlobSyntaxError filepath NoExtensionOnStar =
+     "invalid file glob '" ++ filepath
+  ++ "'. If a wildcard '*' is used it must be with an file extension."
+explainGlobSyntaxError filepath LiteralFileNameGlobStar =
+     "invalid file glob '" ++ filepath
+  ++ "'. If a wildcard '**' is used as a parent directory, the"
+  ++ " file's base name must be a wildcard '*'."
+explainGlobSyntaxError _ EmptyGlob =
+     "invalid file glob. A glob cannot be the empty string."
+explainGlobSyntaxError filepath VersionDoesNotSupportGlobStar =
+     "invalid file glob '" ++ filepath
+  ++ "'. Using the double-star syntax requires 'cabal-version: 3.0'"
+  ++ " or greater. Alternatively, for compatibility with earlier Cabal"
+  ++ " versions, list the included directories explicitly."
+explainGlobSyntaxError filepath VersionDoesNotSupportGlob =
+     "invalid file glob '" ++ filepath
+  ++ "'. Using star wildcards requires 'cabal-version: >= 1.6'. "
+  ++ "Alternatively if you require compatibility with earlier Cabal "
+  ++ "versions then list all the files explicitly."
+
+data IsRecursive = Recursive | NonRecursive
+
+data GlobPat = PatStem String GlobPat
+             -- ^ A single subdirectory component + remainder.
+             | PatMatch IsRecursive String
+             -- ^ First argument: Is this a @**/*.ext@ pattern?
+             --   Second argument: the extensions to accept.
+             | PatLit FilePath
+             -- ^ Literal file name.
+
+fileGlobMatches :: GlobPat -> FilePath -> Bool
+fileGlobMatches pat = fileGlobMatchesSegments pat . splitDirectories
+
+fileGlobMatchesSegments :: GlobPat -> [FilePath] -> Bool
+fileGlobMatchesSegments _ [] = False
+fileGlobMatchesSegments pat (seg : segs) = case pat of
+  PatStem dir pat' ->
+    dir == seg && fileGlobMatchesSegments pat' segs
+  PatMatch Recursive ext ->
+    ext == takeExtensions (foldl' (flip const) seg segs)
```

Comment:
> `(foldl' (flip const) seg segs)`

IMO `last (seg : segs)` is more readable. Or we can vendor `lastDef` from the `safe` package.